Linux/O_STM/DOXY/man/man3/ULSTER.3
.TH "ULSTER" 3 "Sun Apr 1 2018" "C++ Software transactional Memory" \" -*- nroff -*-
.ad l
.nh
.SH NAME
ULSTER \- 
.SH SYNOPSIS
.br
.PP
.PP
\fC#include <ULSTER\&.h>\fP
.PP
Inherits \fBBANK\fP\&.
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "virtual void \fBcopy\fP (std::shared_ptr< \fBOSTM\fP > to, std::shared_ptr< \fBOSTM\fP > from)"
.br
.RI "\fIcopy function, make deep copy of the object/pointer \fP"
.ti -1c
.RI "virtual int \fBGetAccountNumber\fP () const "
.br
.ti -1c
.RI "virtual std::string \fBGetAddress\fP () const "
.br
.ti -1c
.RI "virtual double \fBGetBalance\fP () const "
.br
.ti -1c
.RI "virtual std::shared_ptr< \fBOSTM\fP > \fBgetBaseCopy\fP (std::shared_ptr< \fBOSTM\fP > object)"
.br
.RI "\fIgetBaseCopy function, make deep copy of the object/pointer and Return a new std::shared_ptr<BANK> type object \fP"
.ti -1c
.RI "virtual std::string \fBGetFirstName\fP () const "
.br
.ti -1c
.RI "virtual std::string \fBGetFullname\fP () const "
.br
.ti -1c
.RI "virtual std::string \fBGetLastName\fP () const "
.br
.ti -1c
.RI "\fBULSTER\fP \fBoperator=\fP (const \fBULSTER\fP &orig)"
.br
.ti -1c
.RI "virtual void \fBSetAccountNumber\fP (int \fBaccountNumber\fP)"
.br
.ti -1c
.RI "virtual void \fBSetAddress\fP (std::string \fBaddress\fP)"
.br
.ti -1c
.RI "virtual void \fBSetBalance\fP (double \fBbalance\fP)"
.br
.ti -1c
.RI "virtual void \fBSetFirstName\fP (std::string \fBfirstName\fP)"
.br
.ti -1c
.RI "virtual void \fBSetFullname\fP (std::string \fBfullname\fP)"
.br
.ti -1c
.RI "virtual void \fBSetLastName\fP (std::string \fBlastName\fP)"
.br
.ti -1c
.RI "virtual void \fBtoString\fP ()"
.br
.RI "\fI_cast, is use to cast bak the std::shared_ptr<OSTM> to the required type \fP"
.ti -1c
.RI "\fBULSTER\fP ()"
.br
.ti -1c
.RI "\fBULSTER\fP (int \fBaccountNumber\fP, double \fBbalance\fP, std::string \fBfirstName\fP, std::string \fBlastName\fP, std::string \fBaddress\fP)"
.br
.ti -1c
.RI "\fBULSTER\fP (std::shared_ptr< \fBBANK\fP > obj, int _version, int _unique_id)"
.br
.ti -1c
.RI "\fBULSTER\fP (const \fBULSTER\fP &orig)"
.br
.ti -1c
.RI "virtual \fB~ULSTER\fP ()"
.br
.in -1c
.SS "Private Attributes"

.in +1c
.ti -1c
.RI "int \fBaccountNumber\fP"
.br
.ti -1c
.RI "std::string \fBaddress\fP"
.br
.ti -1c
.RI "double \fBbalance\fP"
.br
.ti -1c
.RI "std::string \fBfirstName\fP"
.br
.ti -1c
.RI "std::string \fBfullname\fP"
.br
.ti -1c
.RI "std::string \fBlastName\fP"
.br
.in -1c
.SH "Detailed Description"
.PP 
Inherit from \fBBANK\fP 
.PP
Definition at line 19 of file ULSTER\&.h\&.
.SH "Constructor & Destructor Documentation"
.PP 
.SS "ULSTER::ULSTER ()\fC [inline]\fP"
Constructor 
.PP
Definition at line 24 of file ULSTER\&.h\&.
.PP
References accountNumber, address, balance, firstName, fullname, and lastName\&.
.PP
Referenced by getBaseCopy(), and ULSTER()\&.
.PP
.nf
24              : BANK() {
25         this->accountNumber = 0;
26         this->balance = 50;
27         this->firstName = "Joe";
28         this->lastName = "Blog";
29         this->address = "High street, Carlow";
30         this->fullname = firstName + " " + lastName;
31     };
.fi
.SS "ULSTER::ULSTER (int accountNumber, double balance, std::string firstName, std::string lastName, std::string address)\fC [inline]\fP"
Custom constructor 
.PP
Definition at line 35 of file ULSTER\&.h\&.
.PP
References accountNumber, address, balance, firstName, fullname, and lastName\&.
.PP
.nf
35                                                                                                           : BANK() {
36         this->accountNumber = accountNumber;
37         this->balance = balance;
38         this->firstName = firstName;
39         this->lastName = lastName;
40         this->address = address;
41         this->fullname = firstName + " " + lastName;
42     };
.fi
.SS "ULSTER::ULSTER (std::shared_ptr< \fBBANK\fP > obj, int _version, int _unique_id)\fC [inline]\fP"
Custom constructor, used by the library for deep copying 
.PP
Definition at line 46 of file ULSTER\&.h\&.
.PP
References accountNumber, address, balance, firstName, fullname, lastName, and ULSTER()\&.
.PP
.nf
46                                                                   : BANK(_version, _unique_id) {
47 
48         this->accountNumber = obj->GetAccountNumber();
49         this->balance = obj->GetBalance();
50         this->firstName = obj->GetFirstName();
51         this->lastName = obj->GetLastName();
52         this->address = obj->GetAddress();
53         this->fullname = obj->GetFirstName() + " " + obj->GetLastName(); 
54     };
.fi
.SS "ULSTER::ULSTER (const \fBULSTER\fP & orig)"
Copy constructor 
.PP
Definition at line 15 of file ULSTER\&.cpp\&.
.PP
.nf
15                                  {
16 }
.fi
.SS "ULSTER::~ULSTER ()\fC [virtual]\fP"
de-constructor 
.PP
Definition at line 18 of file ULSTER\&.cpp\&.
.PP
Referenced by operator=()\&.
.PP
.nf
18                 {
19 }
.fi
.SH "Member Function Documentation"
.PP 
.SS "void ULSTER::copy (std::shared_ptr< \fBOSTM\fP > to, std::shared_ptr< \fBOSTM\fP > from)\fC [virtual]\fP"

.PP
copy function, make deep copy of the object/pointer 
.PP
\fBParameters:\fP
.RS 4
\fIobjTO\fP is a std::shared_ptr<BANK> type object casted back from std::shared_ptr<OSTM> 
.br
\fIobjFROM\fP is a std::shared_ptr<BANK> type object casted back from std::shared_ptr<OSTM> 
.RE
.PP

.PP
Reimplemented from \fBOSTM\fP\&.
.PP
Definition at line 37 of file ULSTER\&.cpp\&.
.PP
References OSTM::Set_Unique_ID()\&.
.PP
Referenced by operator=()\&.
.PP
.nf
37                                                                  {
38 
39     std::shared_ptr<ULSTER> objTO = std::dynamic_pointer_cast<ULSTER>(to);
40     std::shared_ptr<ULSTER> objFROM = std::dynamic_pointer_cast<ULSTER>(from);
41     objTO->Set_Unique_ID(objFROM->Get_Unique_ID());
42     objTO->Set_Version(objFROM->Get_Version());
43     objTO->SetAccountNumber(objFROM->GetAccountNumber());
44     objTO->SetBalance(objFROM->GetBalance());
45 
46     
47 }
.fi
.SS "int ULSTER::GetAccountNumber () const\fC [virtual]\fP"

.PP
Reimplemented from \fBBANK\fP\&.
.PP
Definition at line 83 of file ULSTER\&.cpp\&.
.PP
References accountNumber\&.
.PP
Referenced by operator=(), and toString()\&.
.PP
.nf
83                                    {
84     return accountNumber;
85 }
.fi
.SS "std::string ULSTER::GetAddress () const\fC [virtual]\fP"

.PP
Reimplemented from \fBBANK\fP\&.
.PP
Definition at line 67 of file ULSTER\&.cpp\&.
.PP
References address\&.
.PP
Referenced by operator=()\&.
.PP
.nf
67                                    {
68     return address;
69 }
.fi
.SS "double ULSTER::GetBalance () const\fC [virtual]\fP"

.PP
Reimplemented from \fBBANK\fP\&.
.PP
Definition at line 75 of file ULSTER\&.cpp\&.
.PP
References balance\&.
.PP
Referenced by operator=(), and toString()\&.
.PP
.nf
75                                 {
76     return balance;
77 }
.fi
.SS "std::shared_ptr< \fBOSTM\fP > ULSTER::getBaseCopy (std::shared_ptr< \fBOSTM\fP > object)\fC [virtual]\fP"

.PP
getBaseCopy function, make deep copy of the object/pointer and Return a new std::shared_ptr<BANK> type object 
.PP
\fBParameters:\fP
.RS 4
\fIobjTO\fP is a \fBBANK\fP type pointer for casting 
.br
\fIobj\fP is a std::shared_ptr<BANK> return type 
.RE
.PP

.PP
Reimplemented from \fBOSTM\fP\&.
.PP
Definition at line 25 of file ULSTER\&.cpp\&.
.PP
References ULSTER()\&.
.PP
Referenced by operator=()\&.
.PP
.nf
26 {
27     std::shared_ptr<BANK> objTO = std::dynamic_pointer_cast<BANK>(object);
28     std::shared_ptr<BANK> obj(new ULSTER(objTO,object->Get_Version(),object->Get_Unique_ID())); 
29     std::shared_ptr<OSTM> ostm_obj = std::dynamic_pointer_cast<OSTM>(obj);                                  
30     return ostm_obj;
31 }
.fi
.SS "std::string ULSTER::GetFirstName () const\fC [virtual]\fP"

.PP
Reimplemented from \fBBANK\fP\&.
.PP
Definition at line 99 of file ULSTER\&.cpp\&.
.PP
References firstName\&.
.PP
Referenced by operator=(), and toString()\&.
.PP
.nf
99                                      {
100     return firstName;
101 }
.fi
.SS "std::string ULSTER::GetFullname () const\fC [virtual]\fP"

.PP
Reimplemented from \fBBANK\fP\&.
.PP
Definition at line 107 of file ULSTER\&.cpp\&.
.PP
References fullname\&.
.PP
Referenced by operator=()\&.
.PP
.nf
107                                     {
108     return fullname;
109 }
.fi
.SS "std::string ULSTER::GetLastName () const\fC [virtual]\fP"

.PP
Reimplemented from \fBBANK\fP\&.
.PP
Definition at line 91 of file ULSTER\&.cpp\&.
.PP
References lastName\&.
.PP
Referenced by operator=(), and toString()\&.
.PP
.nf
91                                     {
92     return lastName;
93 }
.fi
.SS "\fBULSTER\fP ULSTER::operator= (const \fBULSTER\fP & orig)\fC [inline]\fP"
Operator 
.PP
Definition at line 62 of file ULSTER\&.h\&.
.PP
References accountNumber, address, balance, copy(), firstName, fullname, GetAccountNumber(), GetAddress(), GetBalance(), getBaseCopy(), GetFirstName(), GetFullname(), GetLastName(), lastName, SetAccountNumber(), SetAddress(), SetBalance(), SetFirstName(), SetFullname(), SetLastName(), toString(), and ~ULSTER()\&.
.PP
.nf
62 {};
.fi
.SS "void ULSTER::SetAccountNumber (int accountNumber)\fC [virtual]\fP"

.PP
Reimplemented from \fBBANK\fP\&.
.PP
Definition at line 79 of file ULSTER\&.cpp\&.
.PP
References accountNumber\&.
.PP
Referenced by operator=()\&.
.PP
.nf
79                                                {
80     this->accountNumber = accountNumber;
81 }
.fi
.SS "void ULSTER::SetAddress (std::string address)\fC [virtual]\fP"

.PP
Reimplemented from \fBBANK\fP\&.
.PP
Definition at line 63 of file ULSTER\&.cpp\&.
.PP
References address\&.
.PP
Referenced by operator=()\&.
.PP
.nf
63                                          {
64     this->address = address;
65 }
.fi
.SS "void ULSTER::SetBalance (double balance)\fC [virtual]\fP"

.PP
Reimplemented from \fBBANK\fP\&.
.PP
Definition at line 71 of file ULSTER\&.cpp\&.
.PP
References balance\&.
.PP
Referenced by operator=()\&.
.PP
.nf
71                                       {
72     this->balance = balance;
73 }
.fi
.SS "void ULSTER::SetFirstName (std::string firstName)\fC [virtual]\fP"

.PP
Reimplemented from \fBBANK\fP\&.
.PP
Definition at line 95 of file ULSTER\&.cpp\&.
.PP
References firstName\&.
.PP
Referenced by operator=()\&.
.PP
.nf
95                                              {
96     this->firstName = firstName;
97 }
.fi
.SS "void ULSTER::SetFullname (std::string fullname)\fC [virtual]\fP"

.PP
Reimplemented from \fBBANK\fP\&.
.PP
Definition at line 103 of file ULSTER\&.cpp\&.
.PP
References fullname\&.
.PP
Referenced by operator=()\&.
.PP
.nf
103                                            {
104     this->fullname = fullname;
105 }
.fi
.SS "void ULSTER::SetLastName (std::string lastName)\fC [virtual]\fP"

.PP
Reimplemented from \fBBANK\fP\&.
.PP
Definition at line 87 of file ULSTER\&.cpp\&.
.PP
References lastName\&.
.PP
Referenced by operator=()\&.
.PP
.nf
87                                            {
88     this->lastName = lastName;
89 }
.fi
.SS "void ULSTER::toString ()\fC [virtual]\fP"

.PP
_cast, is use to cast bak the std::shared_ptr<OSTM> to the required type toString function, displays the object values in formatted way 
.PP
Reimplemented from \fBOSTM\fP\&.
.PP
Definition at line 58 of file ULSTER\&.cpp\&.
.PP
References OSTM::Get_Unique_ID(), OSTM::Get_Version(), GetAccountNumber(), GetBalance(), GetFirstName(), and GetLastName()\&.
.PP
Referenced by operator=()\&.
.PP
.nf
59 {
60    std::cout << "\nULSTER BANK" << "\nUnique ID : " << this->Get_Unique_ID() << "\nInt account : " << this->GetAccountNumber() << "\nDouble value : " << this->GetBalance() << "\nFirst name: " << this->GetFirstName() << "\nLast name : " << this->GetLastName()  << "\nVersion number : " << this->Get_Version() << std::endl;
61 }
.fi
.SH "Member Data Documentation"
.PP 
.SS "int ULSTER::accountNumber\fC [private]\fP"

.PP
Definition at line 95 of file ULSTER\&.h\&.
.PP
Referenced by GetAccountNumber(), operator=(), SetAccountNumber(), and ULSTER()\&.
.SS "std::string ULSTER::address\fC [private]\fP"

.PP
Definition at line 97 of file ULSTER\&.h\&.
.PP
Referenced by GetAddress(), operator=(), SetAddress(), and ULSTER()\&.
.SS "double ULSTER::balance\fC [private]\fP"

.PP
Definition at line 96 of file ULSTER\&.h\&.
.PP
Referenced by GetBalance(), operator=(), SetBalance(), and ULSTER()\&.
.SS "std::string ULSTER::firstName\fC [private]\fP"

.PP
Definition at line 93 of file ULSTER\&.h\&.
.PP
Referenced by GetFirstName(), operator=(), SetFirstName(), and ULSTER()\&.
.SS "std::string ULSTER::fullname\fC [private]\fP"

.PP
Definition at line 92 of file ULSTER\&.h\&.
.PP
Referenced by GetFullname(), operator=(), SetFullname(), and ULSTER()\&.
.SS "std::string ULSTER::lastName\fC [private]\fP"

.PP
Definition at line 94 of file ULSTER\&.h\&.
.PP
Referenced by GetLastName(), operator=(), SetLastName(), and ULSTER()\&.

.SH "Author"
.PP 
Generated automatically by Doxygen for C++ Software transactional Memory from the source code\&.


Client application Tests for Linux/Test05/ULSTER.cpp

/* 
 * File:   ULSTER.cpp
 * Author: Zoltan Fuzesi
 * IT Carlow : C00197361
 *
 * Created on January 17, 2018, 8:02 PM
 */

#include "ULSTER.h"

//ULSTER::ULSTER() {
//}

ULSTER::ULSTER(const ULSTER& orig) {
}

ULSTER::~ULSTER() {
}
/*!
 * \brief getBaseCopy function, make deep copy of the object/pointer and Return a new std::shared_ptr<BANK> type object
 * \param objTO is a BANK type pointer for casting
 * \param obj is a std::shared_ptr<BANK> return type
 */
std::shared_ptr<OSTM> ULSTER::getBaseCopy(std::shared_ptr<OSTM> object)
{
    std::shared_ptr<BANK> objTO = std::dynamic_pointer_cast<BANK>(object);
    std::shared_ptr<BANK> obj(new ULSTER(objTO,object->Get_Version(),object->Get_Unique_ID())); 
    std::shared_ptr<OSTM> ostm_obj = std::dynamic_pointer_cast<OSTM>(obj);                                  
    return ostm_obj;
}
/*!
 * \brief copy function, make deep copy of the object/pointer
 * \param objTO is a std::shared_ptr<BANK> type object casted back from std::shared_ptr<OSTM>
 * \param objFROM is a std::shared_ptr<BANK> type object casted back from std::shared_ptr<OSTM>
 */ 
void ULSTER::copy(std::shared_ptr<OSTM> to, std::shared_ptr<OSTM> from){

	std::shared_ptr<ULSTER> objTO = std::dynamic_pointer_cast<ULSTER>(to);
	std::shared_ptr<ULSTER> objFROM = std::dynamic_pointer_cast<ULSTER>(from);
	objTO->Set_Unique_ID(objFROM->Get_Unique_ID());
	objTO->Set_Version(objFROM->Get_Version());
	objTO->SetAccountNumber(objFROM->GetAccountNumber());
	objTO->SetBalance(objFROM->GetBalance());

    
}
/*!
 * \brief _cast, is use to cast bak the std::shared_ptr<OSTM> to the required type
 */
//std::shared_ptr<ULSTER> ULSTER::_cast(std::shared_ptr<OSTM> _object){
//
//    return static_cast<std::shared_ptr<ULSTER>>(_object);
//}
/*!
 *  \brief toString function, displays the object values in formatted way
 */
void ULSTER::toString()
{
   std::cout << "\nULSTER BANK" << "\nUnique ID : " << this->Get_Unique_ID() << "\nInt account : " << this->GetAccountNumber() << "\nDouble value : " << this->GetBalance() << "\nFirst name: " << this->GetFirstName() << "\nLast name : " << this->GetLastName()  << "\nVersion number : " << this->Get_Version() << std::endl;
}

void ULSTER::SetAddress(std::string address) {
    this->address = address;
}

std::string ULSTER::GetAddress() const {
    return address;
}

void ULSTER::SetBalance(double balance) {
    this->balance = balance;
}

double ULSTER::GetBalance() const {
    return balance;
}

void ULSTER::SetAccountNumber(int accountNumber) {
    this->accountNumber = accountNumber;
}

int ULSTER::GetAccountNumber() const {
    return accountNumber;
}

void ULSTER::SetLastName(std::string lastName) {
    this->lastName = lastName;
}

std::string ULSTER::GetLastName() const {
    return lastName;
}

void ULSTER::SetFirstName(std::string firstName) {
    this->firstName = firstName;
}

std::string ULSTER::GetFirstName() const {
    return firstName;
}

void ULSTER::SetFullname(std::string fullname) {
    this->fullname = fullname;
}

std::string ULSTER::GetFullname() const {
    return fullname;
}

